Code/RDGeneral/RDValue-doublemagic.h

```diff
@@ -0,0 +1,410 @@
+//       nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+#ifndef RDKIT_RDVALUE_PTRMAGIC_H
+#define RDKIT_RDVALUE_PTRMAGIC_H
+
+#include <stdint.h>
+#include <cassert>
+#include <boost/any.hpp>
+#include "Invariant.h"
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+#include <vector>
+#include <string>
+#include <boost/utility.hpp>
+#include <boost/lexical_cast.hpp>
+#include <cmath>
+#include <boost/type_traits.hpp>
+#include <boost/static_assert.hpp>
+#include "LocaleSwitcher.h"
+#include <boost/detail/endian.hpp>
+
+#if defined(BOOST_BIG_ENDIAN) 
+BOOST_STATIC_ASSERT(false && "RDValue only has been tested under little endian cpus");
+#endif
+
+namespace RDKit {
+
+  // Inspired by
+  // https://nikic.github.io/2012/02/02/Pointer-magic-for-efficient-dynamic-value-representations.html
+// 16 bit storage for value types using Quiet NaN spaces in
+//  doubles
+// Won't work on Solaris and some other os's as mmaping maps from
+// top memory down
+// Example check:
+//     std::string *pointer = new std::string(v);
+//      assert((reinterpret_cast<uint64_t>(pointer) & StringTag) == 0);
+
+//  implementations, need a typedef at compile time to figure this out.
+//  current implementation is probably little endian, need to check.
+  
+/*
+  Encoding for storing other things as a double.  Use
+  Quiet NaN
+  Quiet NaN: // used to encode types
+   F   F    F   1XXX < - X = type bits (first bit is set to one)
+
+  seeeeeee|eeeemmmm|mmmmmmmm|mmmmmmmm|mmmmmmmm|mmmmmmmm|mmmmmmmm|mmmmmmmm
+  s1111111|11111ppp|pppppppp|pppppppp|pppppppp|pppppppp|pppppppp|pppppppp
+               ^- first mantissa bit 1    everything else is "payload" -^
+   ^- exponent bits all 1                 and mustn't be all-zero (as it
+  ^- any sign bit                         would be INF then)
+
+  Available 
+  8 = 1000 MaxDouble // Not really a tag, is a sentinel
+  9 = 1001 Float
+  b = 1010 Int32
+  a = 1011 Uint32
+  C = 1100 <none>
+  D = 1101 <none>
+  E = 1110 <none>
+  F = 1111 PtrTag (look at lower 3 bits for type)
+*/ 
+
+namespace RDTypeTag {
```

Comment:
Indeed!  It took me a few tries to get it working as well.  Note that this isn't run by default, you have to use the cmake flag to enable it.

----
Brian Kelley

> On May 9, 2016, at 7:39 AM, Greg Landrum <notifications@github.com> wrote:
> 
> In Code/RDGeneral/RDValue-doublemagic.h:
> 
> > +               ^- first mantissa bit 1    everything else is "payload" -^
> > +   ^- exponent bits all 1                 and mustn't be all-zero (as it
> > +  ^- any sign bit                         would be INF then)
> > +
> > +  Available 
> > +  8 = 1000 MaxDouble // Not really a tag, is a sentinel
> > +  9 = 1001 Float
> > +  b = 1010 Int32
> > +  a = 1011 Uint32
> > +  C = 1100 <none>
> > +  D = 1101 <none>
> > +  E = 1110 <none>
> > +  F = 1111 PtrTag (look at lower 3 bits for type)
> > +*/ 
> > +
> > +namespace RDTypeTag {
> This stuff is voodoo. ;-)
> I'm not going to pretend that I've truly grokked it or that I've reviewed it particularly carefully. The tests (run on a couple different systems) indicate that it all works though, so I'm going to claim that's ok.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>